In [115]:
## Import necessary Libraries
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

**Import Data**

In [116]:
file_name = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-Coursera/laptop_pricing_dataset_mod2.csv"
df = pd.read_csv(file_name,header = 0)
df.head()

Unnamed: 0.1  Unnamed: 0 Manufacturer  Category  GPU  OS  CPU_core  \
0             0           0         Acer         4    2   1         5   
1             1           1         Dell         3    1   1         3   
2             2           2         Dell         3    1   1         7   
3             3           3         Dell         4    2   1         5   
4             4           4           HP         4    2   1         7   

   Screen_Size_inch  CPU_frequency  RAM_GB  Storage_GB_SSD  Weight_pounds  \
0              14.0       0.551724       8             256        3.52800   
1              15.6       0.689655       4             256        4.85100   
2              15.6       0.931034       8             256        4.85100   
3              13.3       0.551724       8             128        2.69010   
4              15.6       0.620690       8             256        4.21155   

   Price Price-binned  Screen-Full_HD  Screen-IPS_panel  
0    978          Low               0                 1  
1    634          Low               1                 0  
2    946          Low               1                 0  
3   1244          Low               0                 1  
4    837          Low               1                 0

In [117]:
df.drop(['Unnamed: 0.1','Unnamed: 0'],axis =1,inplace = True)

In [118]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 238 entries, 0 to 237
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Manufacturer      238 non-null    object 
 1   Category          238 non-null    int64  
 2   GPU               238 non-null    int64  
 3   OS                238 non-null    int64  
 4   CPU_core          238 non-null    int64  
 5   Screen_Size_inch  238 non-null    float64
 6   CPU_frequency     238 non-null    float64
 7   RAM_GB            238 non-null    int64  
 8   Storage_GB_SSD    238 non-null    int64  
 9   Weight_pounds     238 non-null    float64
 10  Price             238 non-null    int64  
 11  Price-binned      238 non-null    object 
 12  Screen-Full_HD    238 non-null    int64  
 13  Screen-IPS_panel  238 non-null    int64  
dtypes: float64(3), int64(9), object(2)
memory usage: 26.2+ KB


In [119]:
df.isnull().sum()

Manufacturer        0
Category            0
GPU                 0
OS                  0
CPU_core            0
Screen_Size_inch    0
CPU_frequency       0
RAM_GB              0
Storage_GB_SSD      0
Weight_pounds       0
Price               0
Price-binned        0
Screen-Full_HD      0
Screen-IPS_panel    0
dtype: int64

**Converting categorical data to numerical data**

In [120]:
df.Manufacturer.value_counts()

Manufacturer
Dell       71
Lenovo     52
HP         49
Acer       19
Asus       18
Toshiba    17
Samsung     5
MSI         4
Huawei      1
Razer       1
Xiaomi      1
Name: count, dtype: int64

In [153]:
from sklearn.preprocessing import LabelEncoder
LE = LabelEncoder()
df['Manufacturer'] = LE.fit_transform(df['Manufacturer'])

In [121]:
df['Price-binned'].value_counts()

Price-binned
Low       160
Medium     72
High        6
Name: count, dtype: int64

In [131]:
df['Price-binned'].unique()

array(['Low', 'Medium', 'High'], dtype=object)

In [149]:
df['Price-binned'] = df['Price-binned'].map({'Low':0, 'Medium':1, 'High':2})

In [155]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 238 entries, 0 to 237
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Manufacturer      238 non-null    int32  
 1   Category          238 non-null    int64  
 2   GPU               238 non-null    int64  
 3   OS                238 non-null    int64  
 4   CPU_core          238 non-null    int64  
 5   Screen_Size_inch  238 non-null    float64
 6   CPU_frequency     238 non-null    float64
 7   RAM_GB            238 non-null    int64  
 8   Storage_GB_SSD    238 non-null    int64  
 9   Weight_pounds     238 non-null    float64
 10  Price             238 non-null    int64  
 11  Price-binned      238 non-null    int64  
 12  Screen-Full_HD    238 non-null    int64  
 13  Screen-IPS_panel  238 non-null    int64  
dtypes: float64(3), int32(1), int64(10)
memory usage: 25.2 KB


In [227]:
final_res = []

In [157]:
X = df.drop('Price',axis=1)
y = df.Price

In [169]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(X,y,test_size = 0.2,random_state = 42)
print(x_train.shape,x_test.shape,y_train.shape,y_test.shape)

(190, 13) (48, 13) (190,) (48,)


**Data Analysis**

In [171]:
from sklearn.linear_model import LinearRegression
LR = LinearRegression()
LR.fit(x_train,y_train)

LinearRegression()

In [173]:
y_pred = LR.predict(x_test)

In [229]:
from sklearn.metrics import r2_score
r2_score_LR = r2_score(y_test,y_pred)
r2_score_LR
final_res.append(r2_score_LR)

In [179]:
from sklearn.ensemble import RandomForestRegressor
RF = RandomForestRegressor()
RF.fit(x_train,y_train)

RandomForestRegressor()

In [181]:
y_pred1 = RF.predict(x_test)

In [231]:
r2_score_RF = r2_score(y_test,y_pred1)
r2_score_RF
final_res.append(r2_score_RF)

In [185]:
from sklearn.tree import DecisionTreeRegressor
DT = DecisionTreeRegressor()
DT.fit(x_train,y_train)

DecisionTreeRegressor()

In [187]:
y_pred2 = DT.predict(x_test)

In [233]:
r2_score_DT = r2_score(y_test,y_pred2)
r2_score_DT
final_res.append(r2_score_DT)

In [201]:
from sklearn.neighbors import KNeighborsRegressor
KNN = KNeighborsRegressor()
KNN.fit(x_train,y_train)

KNeighborsRegressor()

In [203]:
y_pred4 = KNN.predict(x_test)

In [241]:
r2_score_KNN = r2_score(y_test,y_pred4)
r2_score_KNN
final_res.append(r2_score_KNN)

In [209]:
from sklearn.ensemble import GradientBoostingRegressor
GB = GradientBoostingRegressor()
GB.fit(x_train,y_train)

GradientBoostingRegressor()

In [211]:
y_pred5 = GB.predict(x_test)

In [243]:
r2_score_GB = r2_score(y_test,y_pred5)
r2_score_GB
final_res.append(r2_score_GB)

**Conclusion**

In [245]:
final = np.array(final_res)
result = final.reshape(-1,1)
column = ['R2_Score']
index = ['Linear Regression','Random Forest Regressor','Decision Tree Regressor','KNeighbors Regressor','GradientBoostingRegressor']
final_result = pd.DataFrame(result,columns = column, index = index)
final_result

R2_Score
Linear Regression          0.744445
Random Forest Regressor    0.815869
Decision Tree Regressor    0.724276
KNeighbors Regressor       0.406040
GradientBoostingRegressor  0.832805

**We can conclude that GradientBoosting Regressor can be used as the best model as it returns with 83% accuracy.**